## 예외란?
- 프로그램이 정상적인 흐름이 중단되고 예외객체가 생성되는것을 의미합니다. Dart에서는 `Exception`과 `Error` 두가지 유형의 예외 객체가 있습니다.
- **Exception**: 개발자가 예상하고 처리할 수 있는 종류의 예외입니다. 예를들면, 파일을 열 때 파일이 존재하지 않는 경우나 네트워크 요청 중 연결 오류가 발생한 경우 등이 해당됩니다.
- **Error**: 프로그램 내부(컴파일)에서 문제가 있을때 발생합니다. 예를들면, 프로그램이 잘못된 메모리에 접근하려고 시도했을 경우 등이 해당됩니다.

## 예외처리란?
- 프로그램을 설계하고, 실행시(runtime) 예외 상황이 발생할 가능성이 있는 코드를 예측하여, 사전에 예외처리가 되도록 조치하는 것을 의미합니다.

## 예외처리방법은?
- 주로 사용되는 방법은 `try-catch-finally` 구문을 사용하는 것입니다.
- 예외가 발생할 수 있는 코드블록을 `try` 블록 안에 넣고, 발생한 예외를 `catch` 블록에서 잡아 처리할 수 있습니다.
- `finally` 블록은 예외여부와 상관없이 항상 실행되고, 주로 자원을 정리하는 코드를 넣습니다.

## 예외처리 예시코드
```dart
try {
 // 에러가 날 것 같은 코드 작성 
} catch (e) {
    // e : 에러의 정보를 담고 있는 객체
} finally {
  // 예외와 상관없이 실행되는 코드
}
```

## 예외 발생시키기
- 특정 상황에서 예외를 직접 발생시킬 필요가 있을 수 있습니다. 이때 `throw`라는 키워드를 사용합니다.
**예시코드**
```dart
void checkPositive(int number) {
  if (number < 0) {
    throw FormatException('양수가 아닙니다: $number');
  }
}
```

## 정리하기
>예외 처리는 어플리케이션을 안정적으로 유지하고 예상치 못한 상황에 대응하는데 아주 중요한 기능입니다. `try-catch-finally` 구문을 사용하여 예외를 처리하고, 필요한 경우 `throw`를 사용할 수도 있습니다.
